packages/run-protocol/src/vaultFactory/vault.js

```diff
@@ -135,504 +142,553 @@ const initState = (zcf, manager, idInManager) => {
+/**
+ * Calculate the fee, the amount to mint and the resulting debt.
+ * The give and the want together reflect a delta, where typically
+ * one is zero because they come from the gave/want of an offer
+ * proposal. If the `want` is zero, the `fee` will also be zero,
+ * so the simple math works.
+ *
+ * @param {MethodContext} context
+ * @param {Amount} currentDebt
+ * @param {Amount} giveAmount
+ * @param {Amount} wantAmount
+ */
+const loanFee = ({ state }, currentDebt, giveAmount, wantAmount) => {
+  const fee = ceilMultiplyBy(wantAmount, state.manager.getLoanFee());
+  const toMint = AmountMath.add(wantAmount, fee);
+  const newDebt = addSubtract(currentDebt, toMint, giveAmount);
+  return { newDebt, toMint, fee };
+};
 
-    const newCollateralPre = addSubtract(collateralPre, giveColl, wantColl);
-    // max debt supported by current Collateral as modified by proposal
-    const maxDebtPre = await manager.maxDebtFor(newCollateralPre);
+/**
+ * Check whether we can proceed with an `adjustBalances`.
+ *
+ * @param {Amount} newCollateralPre
+ * @param {Amount} maxDebtPre
+ * @param {Amount} newCollateral
+ * @param {Amount} newDebt
+ * @returns {boolean}
+ */
+const checkRestart = (newCollateralPre, maxDebtPre, newCollateral, newDebt) => {
+  if (AmountMath.isGTE(newCollateralPre, newCollateral)) {
+    // The collateral did not go up. If the collateral decreased, we pro-rate maxDebt.
+    // We can pro-rate maxDebt because the quote is either linear (price is
+    // unchanging) or super-linear (also called "convex"). Super-linear is from
+    // AMMs: selling less collateral would mean an even smaller price impact, so
+    // this is a conservative choice.
+    const debtPerCollateral = makeRatioFromAmounts(
+      maxDebtPre,
+      newCollateralPre,
+    );
+    // `floorMultiply` because the debt ceiling should be tight
+    const maxDebtAfter = floorMultiplyBy(newCollateral, debtPerCollateral);
     assert(
-      updaterPre === state.outerUpdater,
-      X`Transfer during vault adjustment`,
+      AmountMath.isGTE(maxDebtAfter, newDebt),
+      X`The requested debt ${q(
+        newDebt,
+      )} is more than the collateralization ratio allows: ${q(maxDebtAfter)}`,
     );
-    assertActive();
+    // The `collateralAfter` can still cover the `newDebt`, so don't restart.
+    return false;
+  }
+  // The collateral went up. Restart if the debt *also* went up because
+  // the price quote might not apply at the higher numbers.
+  return !AmountMath.isGTE(maxDebtPre, newDebt);
+};
 
-    // After the `await`, we retrieve the vault's allocations again,
-    // so we can compare to the debt limit based on the new values.
-    const collateral = getCollateralAllocated(vaultSeat);
-    const newCollateral = addSubtract(collateral, giveColl, wantColl);
+/**
+ * Adjust principal and collateral (atomically for offer safety)
+ *
+ * @param {MethodContext} context
+ * @param {ZCFSeat} clientSeat
+ */
```

Comment:
Needs a return type to prevent circular inference.
```suggestion
  * @returns {string} success message
  */
```

Reply:
Turns out tsc wanted `Promise<string>` rather than `string`.  I _think_ this is right...